 ## Discussion 10A, Spring 2022
 ### Towards Upper-Triangularization By An Orthonormal Basis

Welcome! The discussion 10A worksheet presents arguments that are largely symbolic, but it can be helpful to ground the analysis in a numerical example. 

Run the code cells below for each part to follow along with the given example matrix. This square matrix is referred to as $M$ in the worksheet.

In [1]:
import numpy as np
import scipy.linalg

In [2]:
### defining the given S matrix
M = np.array([
        [5/12, 5/12, 1/6],
        [5/12, 5/12, 1/6],
        [1/6, 1/6, 2/3]
])

print("M is: \n\n", M)

M is: 

 [[0.41666667 0.41666667 0.16666667]
 [0.41666667 0.41666667 0.16666667]
 [0.16666667 0.16666667 0.66666667]]


## [Part (a)] Computing the Basis $U$

In [3]:
### defining a function to perform gram-schmidt on a given vector:
def extend_to_basis(v1: np.ndarray) -> np.ndarray:
    """
    This function uses Gram Schmidt algorithm to extend a n-by-1 column vector v1 to an
    orthonormal basis for R^n
    :param v1: n-by-1 column vector
    :return: n-by-n matrix, orthonormal basis
    """
    ncol = v1.shape[0]
    v1 = np.reshape(v1, (ncol, 1))
    # make matrix vmat with v1 and standard basis vectors
    vmat = np.hstack((v1, np.eye(ncol)))
    # initialize final orthonormal basis matrix
    qmat = np.hstack((v1 / np.linalg.norm(v1), np.zeros((ncol, ncol - 1))))
    idx = 1
    while idx < ncol:
        # Perform Gram Schmidt algorithm
        vvec = vmat[:, idx]
        zvec = vvec.copy()
        for jdx in range(idx):
            zvec -= (vvec.T @ qmat[:, jdx]) * qmat[:, jdx]
        if np.isclose(zvec, qmat[:, idx]).all():
            # this vector is not linearly independent, so delete it
            vmat = np.delete(vmat, idx, 1)
        else:
            qmat[:, idx] = zvec / np.linalg.norm(zvec)
            idx += 1
    return qmat

### calling the given function on our starting vector u, and assigning the result to the matrix U
vec_u = np.array([1, -1, 0])
U = extend_to_basis(vec_u)

print("U is: \n\n", U)

U is: 

 [[ 0.70710678  0.70710678  0.        ]
 [-0.70710678  0.70710678  0.        ]
 [ 0.          0.          1.        ]]


## [Part (b)] Computing $U^\top M U$

In [5]:
### multiplying the matrices in the required order, and rounding to avoid floating point error.
print("U^T M U is: \n\n", np.around(U.T @ M @ U, 4))

### defining R for later subparts
R = U[:,1:]
print("\nthe R matrix is \n\n", R)

U^T M U is: 

 [[0.     0.     0.    ]
 [0.     0.8333 0.2357]
 [0.     0.2357 0.6667]]

the R matrix is 

 [[0.70710678 0.        ]
 [0.70710678 0.        ]
 [0.         1.        ]]


Note that the values here are the decimal equivalent of the fractions in the worksheet solutions. The fractional form is:
$$
U^\top M U =
            \begin{bmatrix}
            	0 & 0 & 0\\
             	0 & \frac{5}{6} & \frac{\sqrt{2}}{6}\\
             	0 & \frac{\sqrt{2}}{6} & \frac{2}{3}
         	\end{bmatrix}
$$

## [Part (c)] Checking that $U^\top = U^{-1}$

In [6]:
### print transpose of U
print("transpose of U: \n\n", U.T)

transpose of U: 

 [[ 0.70710678 -0.70710678  0.        ]
 [ 0.70710678  0.70710678  0.        ]
 [ 0.          0.          1.        ]]


In [7]:
### print inverse of U
print("inverse of U: \n\n", np.linalg.inv(U))

inverse of U: 

 [[ 0.70710678 -0.70710678  0.        ]
 [ 0.70710678  0.70710678  0.        ]
 [ 0.          0.          1.        ]]


Notice the results above are the same!

## [Part (d)] Checking that 
$$ S = U^\top \underbrace{\begin{bmatrix}
            \lambda_1 & \vec{a}^\top \\
            \vec{0} & Q
        \end{bmatrix}}_{T} U $$

In [8]:
### print the left-hand side, which is just S
print("LHS matrix: M is \n\n", np.around(M, 4))

LHS matrix: M is 

 [[0.4167 0.4167 0.1667]
 [0.4167 0.4167 0.1667]
 [0.1667 0.1667 0.6667]]


In [9]:
### print the right-hand side, which is U^T T U, with T defined as above
v1 = np.array([1, -1, 0])/np.sqrt(2)
lambda1 = 0
a = R.T @ M.T @ v1 ### a happens to be zero here, since M was symmetric. Try defining a different M and see what happens!
Q = R.T @ M @ R
print("Q matrix is: \n\n", Q, "\n")

T = np.vstack([
    np.hstack([lambda1, a.T]),
    np.hstack([np.zeros((2, 1)), Q])
])

print("the T matrix is: \n\n", np.around(T, 4), "\n")

print("RHS matrix: \n\n", np.around(U @ T @ U.T, 4))

Q matrix is: 

 [[0.83333333 0.23570226]
 [0.23570226 0.66666667]] 

the T matrix is: 

 [[0.     0.     0.    ]
 [0.     0.8333 0.2357]
 [0.     0.2357 0.6667]] 

RHS matrix: 

 [[0.4167 0.4167 0.1667]
 [0.4167 0.4167 0.1667]
 [0.1667 0.1667 0.6667]]


Note that the RHS and LHS match!

## [Part (e)] Showing that 
$$M = \begin{bmatrix} \vec{v}_1 & R\vec{v}_2 & RY \end{bmatrix}
        \begin{bmatrix} \lambda_1 & a_1 & \vec{a}_\text{rest}^\top \\
            0 & \lambda_2 & \vec{b}^\top    \\
            \vec{0} & 0 & P
        \end{bmatrix}
        \begin{bmatrix} \vec{v}_1 & R\vec{v}_2 & RY \end{bmatrix}^\top$$

In [10]:
### First, we can list the eigenvalues and eigenvectors of Q to choose
np.linalg.eig(Q)

(array([1. , 0.5]),
 array([[ 0.81649658, -0.57735027],
        [ 0.57735027,  0.81649658]]))

We take the pair $$(\lambda_2, \vec{v}_2) = \left(1, \begin{bmatrix} -0.57735027 \\ 0.81649658 \end{bmatrix}\right) = \left(1, \begin{bmatrix} \frac{-\sqrt{3}}{3} \\ \frac{\sqrt{6}}{3} \end{bmatrix}\right)$$

Then, we can make sure that Q actually matches the given form, by defining $Y$ and then carrying out the computation

In [11]:
lambda2 = 0.5
v2 = np.array([[-np.sqrt(3)/3],  [np.sqrt(6)/3]])
Y = np.array([[np.sqrt(6)/3], [np.sqrt(3)/3]])
print("Y, which is the orthonormal extension of v2, is: \n\n", Y)

Y, which is the orthonormal extension of v2, is: 

 [[0.81649658]
 [0.57735027]]


In [12]:
print("Q is: \n\n", Q, "\n")

Q is: 

 [[0.83333333 0.23570226]
 [0.23570226 0.66666667]] 



In [13]:
### first we define the v2, Y concatenated matrix
v2Y = np.hstack([v2, Y])
print("the [v2, Y] matrix is: \n\n", v2Y, "\n")

### Now, we define the P matrix and b vector, analogous to what we did before
P = (Y.T @ Q @ Y).reshape(1,)
b = (Y.T @ Q.T @ v2).reshape(1,)

### Now we define the middle matrix of Q, called T2
T2 = np.vstack([
    np.hstack([lambda2, b.T]),
    np.hstack([0, P])
])

print("The P matrix is: \n\n", np.around(P, 4), "\n")

print("The middle matrix, T2, is: \n\n", np.around(T2, 4), "\n")

print("the RHS of the Q expression is: \n\n", v2Y @ T2 @ v2Y.T, "\n")

the [v2, Y] matrix is: 

 [[-0.57735027  0.81649658]
 [ 0.81649658  0.57735027]] 

The P matrix is: 

 [1.] 

The middle matrix, T2, is: 

 [[0.5 0. ]
 [0.  1. ]] 

the RHS of the Q expression is: 

 [[0.83333333 0.23570226]
 [0.23570226 0.66666667]] 



The left and right sides of the given equation match! Now let's show that our expression for $M$ holds as well:

In [15]:
print("the M matrix is: \n\n", M, "\n")

the M matrix is: 

 [[0.41666667 0.41666667 0.16666667]
 [0.41666667 0.41666667 0.16666667]
 [0.16666667 0.16666667 0.66666667]] 



In [16]:
### we define the block [v1, Rv2, RY]
M_blk = np.hstack([v1.reshape(3,1), R @ v2, R @ Y])

### We form the middle matrix for S
M_mid = np.vstack([
    np.hstack([lambda1, a[0], a[1:].T]),
    np.hstack([0, lambda2, b.T]),
    np.hstack([np.zeros((1,)), 0, P[0]])
])

### We print out the calculated  matrix
print("the RHS matrix is: \n\n", M_blk @ M_mid @ M_blk.T, "\n")

the RHS matrix is: 

 [[0.41666667 0.41666667 0.16666667]
 [0.41666667 0.41666667 0.16666667]
 [0.16666667 0.16666667 0.66666667]] 



## In this section we extend the above algorithm for $n\times n$ matrices

In [18]:
def GramSchmidtRest(v):
    n = v.shape[0]
    v = v / np.linalg.norm(v)
    I = np.identity(n)
    mat = np.hstack((v, I))
    basis = scipy.linalg.orth(mat)
    return basis[:, 1:]

## Recursive formulation from lecture
def UpperTriangularize(M):
    if M.shape[0] == 1:
        return np.array([[1]])
    else:
        eigval, eigvec = scipy.linalg.eig(M)
        v = eigvec[:, :1]
        R = GramSchmidtRest(v)
        B = R.T @ M @ R
        U_1 = UpperTriangularize(B)
        U = np.hstack((v, R@U_1))
        return U

U = UpperTriangularize(M)   

# triangularize matrix M
T = U.T @ M @ U

# print friendly round off
print(np.around(T, 3))

[[ 1.   0.  -0. ]
 [ 0.   0.5 -0. ]
 [-0.  -0.   0. ]]


In [19]:
def uppertriangularizeloop(M):
    currentMatrix = M
    U = np.array([])
    R = np.identity(M.shape[0])
    while currentMatrix.shape[0] > 0:
        eigval, eigvec = scipy.linalg.eig(currentMatrix)
        eigvec = eigvec[:, :1]
        if U.shape[0] == 0:
            U = np.dot(R, eigvec)
        else:
            U = np.hstack((U, np.dot(R, eigvec)))
        if currentMatrix.shape[0] == 1:
            currentMatrix = np.array([])
        else:
            Y = GramSchmidtRest(eigvec)
            currentMatrix = np.dot(Y.T, np.dot(currentMatrix, Y))
            R = np.dot(R, Y)
    return U
        

In [20]:
U = uppertriangularizeloop(M)

# triangularize M
T = U.T @ M @ U

print(np.around(T, 3))


[[ 1.   0.  -0. ]
 [ 0.   0.5 -0. ]
 [-0.  -0.   0. ]]


Wonderful! We have now expressed the original $M$ matrix in an upper-triangular form using orthonormalization